artiq/coredevice/ttl.py

```diff
@@ -206,88 +182,124 @@ def pulse(self, duration):
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_rising(self, duration):
         """Register rising edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(1)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_falling(self, duration):
         """Register falling edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(2)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_both(self, duration):
         """Register both rising and falling edge events for the specified
         duration (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
-    def count(self):
-        """Poll the RTIO input during all the previously programmed gate
-        openings, and returns the number of registered events.
+    def count(self, up_to_timestamp_mu=-1):
+        """Poll the RTIO input up to the specified timestamp, and returns the
+        number of registered events.
 
-        This function does not interact with the time cursor."""
+        If no timestamp is specified, the current cursor position is used. This
+        unnecessarily reduces the slack if the last gate closed further in the
+        past; prefer specifying a timestamp.
+
+        This function does not modify the timeline cursor."""
+        if up_to_timestamp_mu == -1:
+            up_to_timestamp_mu = now_mu()
         count = 0
-        while rtio_input_timestamp(self.i_previous_timestamp, self.channel) >= 0:
+        while rtio_input_timestamp(up_to_timestamp_mu, self.channel) >= 0:
             count += 1
         return count
 
     @kernel
-    def timestamp_mu(self):
+    def timestamp_mu(self, up_to_timestamp_mu=-1):
         """Poll the RTIO input and returns an event timestamp (in machine
```

Comment:
While we are here: Maybe `returns the first event timestamp` to clarify which and add a sentence saying that at most one input event is consumed.